src/clj_irods/core.clj

```diff
@@ -0,0 +1,205 @@
+               (init/init
+                 (:host c)
+                 (:port c)
+                 (:username c)
+                 (:password c)
+                 (:home c)
+                 (:zone c)
+                 (:resource c)
+                 :max-retries (:max-retries c)
+                 :retry-sleep (:retry-sleep c)
+                 :use-trash (:use-trash c))))))
+
+(def icat-spec
+  (memoize (fn [c]
+             (when c
+               (icat-direct/icat-db-spec
+                 (:host c)
+                 (:user c)
+                 (:password c)
+                 :port (:port c)
+                 :db (:db c))))))
+
+(defn have-icat
+  []
+  (map? icat-direct/icat))
+
+(defmacro maybe-jargon
+  [use-jargon jargon-cfg jargon-sym & body]
+  `(if ~use-jargon
+     (init/with-jargon ~jargon-cfg :lazy true [~jargon-sym]
+       (do ~@body))
+     (let [~jargon-sym (delay (throw+ {:type :no-irods}))]
+      (do ~@body))))
+
+(defmacro maybe-icat-transaction
+  [use-icat & body]
+  `(if ~use-icat
+     (icat-direct/with-icat-transaction span-sym#
+       (do ~@body))
+     (do ~@body)))
+
+(defn make-threadpool
+  [prefix thread-count]
+  (let [counter (atom 0)]
+    (Executors/newFixedThreadPool
+      thread-count
+      (proxy [ThreadFactory] []
+        (newThread [^Runnable runnable]
+          (let [t (Thread. runnable)]
+            (.setName t (str prefix "-" (swap! counter inc)))
+            t))))))
+
+(defmacro with-irods
+  "Open connections to iRODS and/or transactions in the ICAT depending on the
+  options passed in `cfg` and what's configured.  Bind this to the symbol
+  passed in as `sym`. Recommended choice is to call it irods."
+  [cfg sym & body]
+  `(let [id# (name (gensym "with-irods-"))
+         jargon-cfg# (jargon-cfg (:jargon ~cfg))
+         use-jargon# (boolean jargon-cfg#)
+         use-icat# (have-icat)
+         jargon-pool# (or (:combined-pool ~cfg) (:jargon-pool ~cfg) (make-threadpool (str id# "-jargon") (or (:jargon-pool-size ~cfg) 1)))
+         icat-pool#   (or (:combined-pool ~cfg) (:icat-pool ~cfg) (make-threadpool (str id# "-icat") (or (:icat-pool-size ~cfg) 5)))]
+     (try+
+       (maybe-icat-transaction use-icat#
+         (maybe-jargon use-jargon# jargon-cfg# jargon#
+           (let [~sym {:jargon      jargon#
+                       :jargon-pool jargon-pool#
+                       :icat-pool   icat-pool#
+                       :has-jargon  use-jargon#
+                       :has-icat    use-icat#
+                       :cache       (atom {})}]
+             (do ~@body))))
+       (finally
+         (when-not (:retain-jargon-pool ~cfg) (.shutdown jargon-pool#))
+         (when-not (:retain-icat-pool ~cfg) (.shutdown icat-pool#))))))
+
+(defn- cached-or-get
+  [irods & get-fns]
+  (or
+    (first (filter delay?
+        (concat
+          (map #(apply (first %) true  irods (rest %)) get-fns)
+          (map #(apply (first %) false irods (rest %)) get-fns))))
+    nil))
```

Comment:
`get-fns` is a set of function-call-like vectors, e.g. `[from-listing extract-fn user zone path]`. In order, they'll be called first with `true` and `irods` put as the first two arguments (`true` for "only use the cache"), and then with `false` instead, halting execution as soon as we get a `delay` rather than the `nil` we get if the data source is disabled or if it's a cache miss.